File: sdks/java/src/main/java/com/backlex/QueryBuilder.java
package com.backlex;

import java.util.Arrays;
import java.util.function.Function;

/** Chainable builder that compiles to a {@link ListQuery} and runs it. */
public final class QueryBuilder<T> {

    private final Function<ListQuery, ListResponse<T>> listFn;
    private final ListQuery q = new ListQuery();

    QueryBuilder(Function<ListQuery, ListResponse<T>> listFn) {
        this.listFn = listFn;
    }

    public QueryBuilder<T> where(Condition cond) {
        q.filter = Filter.normalize(cond);
        return this;
    }

    /** Replace the filter with a raw canonical condition (escape hatch). */
    public QueryBuilder<T> rawFilter(Condition cond) {
        q.filter = Filter.normalize(cond);
        return this;
    }

    public QueryBuilder<T> select(String... fields) {
        q.fields.addAll(Arrays.asList(fields));
        return this;
    }

    public QueryBuilder<T> orderBy(String... sorts) {
        q.sort.addAll(Arrays.asList(sorts));
        return this;
    }

    public QueryBuilder<T> limit(int n) {
        q.limit = n;
        return this;
    }

    public QueryBuilder<T> offset(int n) {
        q.offset = n;
        return this;
    }

    /** Request an extra COUNT: "filter_count", "total_count", or "*". */
    public QueryBuilder<T> withMeta(String m) {
        q.meta = m;
        return this;
    }

    /** The assembled {@link ListQuery} — the canonical input the API takes. */
    public ListQuery toQuery() {
        return q;
    }

    public ListResponse<T> list() {
        return listFn.apply(q);
    }
}

File: sdks/kotlin/src/main/kotlin/com/backlex/QueryBuilder.kt
package com.backlex

/** Chainable builder that compiles to a [ListQuery] and runs it. */
class QueryBuilder<T>(private val listFn: (ListQuery) -> ListResponse<T>) {
    private val q = ListQuery()

    fun where(cond: Condition) = apply { q.filter = Filter.normalize(cond) }

    /** Replace the filter with a raw canonical condition (escape hatch). */
    fun filter(cond: Condition) = apply { q.filter = Filter.normalize(cond) }

    fun select(vararg fields: String) = apply { q.fields.addAll(fields) }

    fun orderBy(vararg sorts: String) = apply { q.sort.addAll(sorts) }

    /** Inline single-hop relations (replaces each FK with the related object). */
    fun expand(vararg rels: String) = apply { q.expand.addAll(rels) }

    /** Project i18n_text fields to one locale, or "*" for the full map. */
    fun locale(loc: String) = apply { q.locale = loc }

    /** Free-text search across readable text fields. */
    fun search(text: String) = apply { q.q = text }

    fun limit(n: Int) = apply { q.limit = n }

    fun offset(n: Int) = apply { q.offset = n }

    /** Request an extra COUNT: "filter_count", "total_count", or "*". */
    fun withMeta(m: String) = apply { q.meta = m }

    /** The assembled [ListQuery] — the canonical input the API takes. */
    fun toQuery(): ListQuery = q

    fun list(): ListResponse<T> = listFn(q)
}
